Add the ability to define a Python based command that uses CommandObjectParsed

jimingham opened the pull request:
This allows you to specify options and arguments and their definitions and then have lldb handle the completions, help, etc. in the same way that lldb does for its parsed commands internally.

This feature has some design considerations as well as the code, so I've also set up an RFC, but I did this one first and will put the RFC address in here once I've pushed it...

Note, the lldb "ParsedCommand interface" doesn't actually do all the work that it should.  For instance, saying the type of an option that has a completer doesn't automatically hook up the completer, and ditto for argument values.  We also do almost no work to verify that the arguments match their definition, or do auto-completion for them.  This patch allows you to make a command that's bug-for-bug compatible with built-in ones, but I didn't want to stall it on getting the auto-command checking to work all the way correctly.

As an overall design note, my primary goal here was to make an interface that worked well in the script language.  For that I needed, for instance, to have a property-based way to get all the option values that were specified.  It was much more convenient to do that by making a fairly bare-bones C interface to define the options and arguments of a command, and set their values, and then wrap that in a Python class (installed along with the other bits of the lldb python module) which you can then derive from to make your new command.  This approach will also make it easier to experiment.

See the file test_commands.py in the test case for examples of how this works.

jimingham committed:
Add the ability to define a Python based command that uses the
CommandObjectParsed way to specify options and arguments so that
these commands can be "first class citizens" with build-in commands.

jimingham commented:
I described this in some more detail here:

https://discourse.llvm.org/t/rfc-parsed-commands-in-python/74532

jimingham commented:
Note, I haven't run this though black or clang-format yet...

bulbazord commented on lldb/source/Commands/CommandObjectCommands.cpp:
```diff
@@ -1255,6 +1258,676 @@ class CommandObjectScriptingObject : public CommandObjectRaw {
+    }
+    
+  private:
+    struct EnumValueStorer {
```
I would say `EnumValueStorage` -- The enum value is in storage, but this isn't doing the storing. Something else is storing it in this struct.

jimingham replied:
Sure

bulbazord commented on lldb/source/Commands/CommandObjectCommands.cpp:
```diff
@@ -1255,6 +1258,676 @@ class CommandObjectScriptingObject : public CommandObjectRaw {
+        element.value = 0;
+      }
+      
+      EnumValueStorer(std::string &in_str_val, std::string &in_usage, 
+          size_t in_value) : value(in_str_val), usage(in_usage) {
```
I recommend passing by value and moving instead of passing by value and copying.

jimingham replied:
Why?

bulbazord replied:
So in general, my (admittedly simplified) understanding of this:
The parameter being a reference type means that when you construct `value` with `value(in_str_val)` you'll end up copying no matter what. It doesn't matter if you passed an r-value to `EnumValueStorer`, you're going to end up constructing 2 `std::string` objects instead of 1.
However, if `in_str_val` was type `std::string` and you performed a move (e.g. `value(std::move(in_str_val))`), then if you pass an r-value to the constructor, you'll end up avoiding a copy (and end up only constructing 1 std::string total). It sounds like a minor optimization, but unnecessary copies are one of those "death by a thousand paper cuts" kind of things, especially if you have lots of tiny extra allocations.

Maybe some of the newer copy elision guarantees from c++17 make this a non-issue? I think I may need to do some more reading...

jimingham replied:
I made the change, but I hope along with your last sentence that this sort of voodoo is not really required to make the call performant.  That's a bit TOO fiddly.

bulbazord commented on lldb/source/Commands/CommandObjectCommands.cpp:
```diff
@@ -1255,6 +1258,676 @@ class CommandObjectScriptingObject : public CommandObjectRaw {
+  };
+
+public:
+  CommandObjectScriptingObjectParsed(CommandInterpreter &interpreter,
+                               std::string name,
+                               StructuredData::GenericSP cmd_obj_sp,
+                               ScriptedCommandSynchronicity synch)
+      : CommandObjectParsed(interpreter, name.c_str()), 
+        m_cmd_obj_sp(cmd_obj_sp), m_synchro(synch), 
+        m_options(interpreter, cmd_obj_sp), m_fetched_help_short(false), 
+        m_fetched_help_long(false) {
```
I think it would be worth creating a static factory method instead of having a fallible constructor for this. If you have no ScriptInterpreter, this fails but you still get a half-baked `CommandObjectScriptingObjectParsed` object.

jimingham replied:
Okay

jimingham replied:
That's CommandObjectScriptingObjectParsed::Create, just above this comment.

bulbazord commented on lldb/source/Interpreter/CommandObject.cpp:
```diff
@@ -447,6 +447,22 @@ bool CommandObject::IsPairType(ArgumentRepetitionType arg_repeat_type) {
 
+std::optional<ArgumentRepetitionType> 
+CommandObject::ArgRepetitionFromString(llvm::StringRef string) {
+  if (string == "plain") return eArgRepeatPlain ;   
+  if (string ==  "optional") return eArgRepeatOptional;
+  if (string ==  "plus") return eArgRepeatPlus;
+  if (string ==  "star") return eArgRepeatStar; 
+  if (string ==  "range") return eArgRepeatRange;
+  if (string ==  "pair-plain") return eArgRepeatPairPlain;
+  if (string ==  "pair-optional") return eArgRepeatPairOptional;
+  if (string ==  "pair-plus") return eArgRepeatPairPlus;
+  if (string ==  "pair-star") return eArgRepeatPairStar;
+  if (string ==  "pair-range") return eArgRepeatPairRange;
+  if (string ==  "pair-range-optional") return eArgRepeatPairRangeOptional;
+  return {};
```
I'd recommend using `llvm::StringSwitch` for something like this.

jimingham replied:
Sure

bulbazord commented on lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp:
```diff
@@ -2755,6 +2755,58 @@ bool ScriptInterpreterPythonImpl::RunScriptBasedCommand(
   return ret_val;
 }
 
+bool ScriptInterpreterPythonImpl::RunScriptBasedParsedCommand(
+    StructuredData::GenericSP impl_obj_sp, Args &args,
+    ScriptedCommandSynchronicity synchronicity,
+    lldb_private::CommandReturnObject &cmd_retobj, Status &error,
```
Why return a bool instead of a `Status` object? You've got an error mutable ref already, that could just be the return value

jimingham replied:
In this patch, I was just adding the Parsed command to the Raw one that was already there, without changing more than I needed to.  If we want to go back and clean up little things like this in both interfaces, I'd rather do that as a separate piece of work.

bulbazord replied:
Makes sense to me.

bulbazord commented on lldb/source/Plugins/ScriptInterpreter/Python/ScriptInterpreterPython.cpp:
```diff
@@ -2873,6 +2925,204 @@ uint32_t ScriptInterpreterPythonImpl::GetFlagsForCommandObject(
+  if (PyErr_Occurred())
+    PyErr_Clear();
+
+  // FIXME: this should really be a void function
```
Why?

jimingham replied:
The function being called is a void return, but I didn't see a way to call that.  I was explaining why I ignored the return, but I made the comment clearer

bulbazord reviewed (commented):
I did an initial pass over your patch but I didn't read the python tests yet

hawkinsw commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,315 @@
+
+The arguments will be in a python array as strings.  
+
+You can access the option values using varname you passed in when defining the option.  
```
```suggestion
You can access the option values using the varname you passed in when defining the option.  
```

hawkinsw commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,315 @@
+        # This makes the ivars for all the varnames in the array and gives them
+        # their default values.
+        for elem in self.options_array:
+            elem['_value_set'] = False       
```
Seeing the hardcoded string literals throughout this function makes me think it might be good to abstract those behind some constants so that they could be changed in the future?

jimingham replied:
These shouldn't be used outside of the parser itself, so I'm not sure that we'd get anything out of one more layer of abstraction here.

hawkinsw commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,315 @@
+    def set_option_value(self, exe_ctx, opt_name, opt_value):
+        return self.get_parser().set_option_value(exe_ctx, opt_name, opt_value)
+
+    # These are the two "pure virtual" methods:
```
Could you decorate them with `@abstractmethod`? https://docs.python.org/3/library/abc.html#abc.abstractmethod

hawkinsw commented:
@jimingham Looks like great work -- I hope my few comments were helpful!

clayborg reviewed (commented):
I like this idea. Since python is so dynamic, I wonder if we can just add extra metadata to the standard "argparse" objects so that users can just write their stuff in the most pythonic way possible using `import argparse` and then adding some extra metadata that lldb can access. For example something like:

```
$ python3
Python 3.8.9 (default, Jul 27 2021, 02:53:04) 
[Clang 7.1.0 (tags/RELEASE_710/final)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import argparse
>>> ap = argparse.ArgumentParser()
>>> v = ap.add_argument('-v')
>>> print(v)
_StoreAction(option_strings=['-v'], dest='v', nargs=None, const=None, default=None, type=None, choices=None, help=None, metavar=None)
>>> print(type(v))
<class 'argparse._StoreAction'>
>>> v.lldb_type = lldb.eArgTypeBoolean
>>> v.lldb_type
lldb.eArgTypeBoolean
```
Then users can use the standard `argparse` module and just add a bit of extra metadata to it that LLDB can then access when needed. This would remove the need for a custom LLDBOVParser class.

medismailben commented on lldb/bindings/python/python-wrapper.swig:
```diff
@@ -831,6 +831,37 @@ bool lldb_private::python::SWIGBridge::LLDBSwigPythonCallCommandObject(
 }
 
+bool lldb_private::python::SWIGBridge::LLDBSwigPythonCallParsedCommandObject(
+    PyObject *implementor, lldb::DebuggerSP debugger, lldb_private::StructuredDataImpl &args_impl,
```
I'll get rid of function altogether later this week.

jimingham replied:
It's spelled incorrectly in the whole rest of this file except for the thread plan instances.  We should either get rid of all of these by using ScriptedPythonInterface, or fix all the spellings to be consistent, but that seems better as a separate commit.

jimingham replied:
Note, whoever spelt this originally was sure implementor was what they wanted to use, the C++ files use the same spelling.

medismailben commented on lldb/bindings/python/python-wrapper.swig:
```diff
@@ -831,6 +831,37 @@ bool lldb_private::python::SWIGBridge::LLDBSwigPythonCallCommandObject(
+  PyErr_Cleaner py_err_cleaner(true);
+
+  PythonObject self(PyRefType::Borrowed, implementor);
+  auto pfunc = self.ResolveName<PythonCallable>("__call__");
```
ditto.

jimingham replied:
We don't do any checking for arguments in any of the `pfunc` lookups in the python wrapper, except in one or two cases where we support overloaded implementation functions.  I don't think doing this piecemeal is a great idea.  It would be better to have some kind of `check_pfunc` feature that does the lookup and takes some specification about what to expect of the found function.  Then we can insert this in all the current lookups.
But this patch is long enough already, it would be better to do that as a separate patch.

medismailben commented on lldb/bindings/python/python-wrapper.swig:
```diff
@@ -831,6 +831,37 @@ bool lldb_private::python::SWIGBridge::LLDBSwigPythonCallCommandObject(
+  PythonObject self(PyRefType::Borrowed, implementor);
+  auto pfunc = self.ResolveName<PythonCallable>("__call__");
+
+  if (!pfunc.IsAllocated())
+    return false;
```
ditto.

jimingham replied:
I added this here, but we really need a more general mechanism for informing the user when the python code we've been handed is lacking some affordance.  Most of these wrappers don't have a CommandReturnObject to complain through.

medismailben commented on lldb/bindings/python/python-wrapper.swig:
```diff
@@ -831,6 +831,37 @@ bool lldb_private::python::SWIGBridge::LLDBSwigPythonCallCommandObject(
+  if (!pfunc.IsAllocated())
+    return false;
+
+  auto cmd_retobj_arg = SWIGBridge::ToSWIGWrapper(cmd_retobj);
```
ditto.

jimingham replied:
That must have been problematic in development.  This sort of compressed style is less good when you are debugging...

medismailben commented on lldb/bindings/python/python-wrapper.swig:
```diff
@@ -831,6 +831,37 @@ bool lldb_private::python::SWIGBridge::LLDBSwigPythonCallCommandObject(
+
+  auto cmd_retobj_arg = SWIGBridge::ToSWIGWrapper(cmd_retobj);
+
+  // FIXME:
+  // I wanted to do something like:
+  // size_t num_elem = args.size();
+  // PythonList my_list(num_elem);
+  // for (const char *elem : args)
+  //   my_list.append(PythonString(elem);
+  //
+  // and then pass my_list to the pfunc, but that crashes somewhere
+  // deep in Python for reasons that aren't clear to me.
```
My guess is that the backtrace only contains for python internal non symbolicated frames that are basically useless for debugging. I've experienced that myself multiple times.

jimingham commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,315 @@
+"""
+This module implements a couple of utility classes to make writing
+lldb parsed commands more Pythonic.
+The way to use it is to make a class for you command that inherits from ParsedCommandBase.
```
The problem with that approach is that completion happens in C++ in the CommandInterpreter, and if argparse were doing the parsing, to support completion the CommandInterpreter would have to call back to argparse to figure out which option was being completed.  
What I tried to do is make the interface for defining your options look the same as what you would have written for argparse so that if you have an extant command it will be easy to convert it and if you know argparse, you know how to define lldb commands already, the only difference being you don't use argparse, you add it to our object instead.

Look at the test_commands.py in the tests to get a sense of what that looks like.

jimingham replied:
I converted the lldb/examples/python/cmdtemplate.py to the new form in the latest patch, if you want to get a sense of what you have to do to convert from one form to the other.  I think this is pretty straightforward.  But it also can't be a straight translation because you need to be able to specify the lldb.eArgType... .  Anyway, have a look at the diff between the two, and if you can think of anything to make this easier, I'm happy for suggestions (maybe as a follow-on patch, if this looks roughly okay, however...)

jimingham commented:
I actually run against a hand-build Python with debug symbols, so it was in fact a backtrace that contains only python internal functions, and even though they had symbols I still couldn't tell what was going on...  There must be a bunch of handy tricks for debugging the Python interpreter, but I don't know them.

Jim


> On Nov 1, 2023, at 11:06 PM, Med Ismail Bennani ***@***.***> wrote:
> 
> 
> @medismailben commented on this pull request.
> 
> In lldb/bindings/python/python-wrapper.swig <https://github.com/llvm/llvm-project/pull/70734#discussion_r1379633468>:
> 
> > +  // FIXME:
> +  // I wanted to do something like:
> +  // size_t num_elem = args.size();
> +  // PythonList my_list(num_elem);
> +  // for (const char *elem : args)
> +  //   my_list.append(PythonString(elem);
> +  //
> +  // and then pass my_list to the pfunc, but that crashes somewhere
> +  // deep in Python for reasons that aren't clear to me.
> My guess is that the backtrace only contains for python internal non symbolicated frames that are basically useless for debugging. I've experienced that myself multiple times.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/llvm/llvm-project/pull/70734#discussion_r1379633468>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVW4NABS6EBDGMGSXDE3YCMZ6HAVCNFSM6AAAAAA6WXATNCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTOMBZGUYTCNRTGE>.
> You are receiving this because you were mentioned.
>

jimingham commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,315 @@
+    # Some methods to translate common value types.  Should return a
+    # tuple of the value and an error value (True => error) if the
+    # type can't be converted.
+    # FIXME: Need a way to push the conversion error string back to lldb.
```
Do we raise Python exceptions anywhere else in our API's.  It is a pythonic way to do things, but we don't tend to do it that way.

bulbazord replied:
We don't do it often but there is prior art. Specifically in a few of our SBAPI extensions, such as SBAddress and SBData. My $0.02 is that it's worth making these interfaces more pythonic since people will want to write their own commands with python.

jimingham replied:
I still think exceptions would be an inconvenient way to do this.  These aren't API's that users are going to call directly, so we aren't matching expectations for Python function behavior.  Rather, in the middle of parsing a command from lldb (happening in C++) the command parser will arrange to have this called.  So if it raises a Python exception, the LLDB SWIG wrappers will have to catch that exception and turn it into a Status with the error we want to report.  I think it would be much less fragile to just pipe an SBError through directly and use that for error reporting.

jimingham commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,315 @@
+
+llvm-project/lldb/test/API/commands/command/script/add/test_commands.py
+
+FIXME: I should make a convenient wrapper for that. 
```
Comment got displaced, it goes with the "was_set" discussion.

jimingham commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,315 @@
+            return cls.translators[value_type](value)
+        except KeyError:
+            # If we don't have a translator, return the string value.
+            return (value, False)
```
The translator methods return `(Value, Error)` if there's no translator then there's no error, so I have to return False here as well.

bulbazord replied:
I think I missed that the first time I read this. Thanks for clarifying.

I wonder if it might be worth introducing a `NamedTuple` type representing this if it's going to be the interface we go with?

jimingham commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,315 @@
+            # If we don't have a translator, return the string value.
+            return (value, False)
+
+    # FIXME: would this be better done on the C++ side?
+    # The common completers are missing some useful ones.
+    # For instance there really should be a common Type completer
+    # And an "lldb command name" completer.
+    completion_table = {
```
I agree as well, but I don't want to do that sort of change in this patch.

bulbazord replied:
Is doing it in C++ independent of this change? I think it would be worth writing more about this (either in the FIXME or in a GitHub issue). I've been slowly learning and understanding how CommandObjects work but I have very little knowledge on the completer machinery.

jimingham replied:
The C++ side of the option specification needs work.  When you actually define an option on the C++ side, you give it an eArgType, but that doesn't actually wire up completions directly.  Instead, you have to also do that by hand.  So writing the C++ based commands in inconvenient, and would make writing ones in Python tedious as well.  I don't want people to have to get in the habit of manually doing that, so I am just providing this shim.  The change to get the ArgType -> completer working correctly is much more work, and mostly orthogonal to the main purpose of this patch.

jimingham commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,315 @@
+                   value_type=lldb.eArgTypeNone, completion_type=None,
+                   enum_values=None):
+        """
+        short_option: one character, must be unique, not required
+        long_option: no spaces, must be unique, required
```
The actual command definition on the C++ side will check for uniqueness and return an error.  I don't think we need to do that in two places.

jimingham commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,315 @@
+    @staticmethod
+    def do_register_cmd(cls, debugger, module_name):
+        # Add any commands contained in this module to LLDB
+        command = "command script add -o -p -c %s.%s %s" % (
+            module_name,
+            cls.__name__,
+            cls.program,
+        )
+        debugger.HandleCommand(command)
+        print(
+            'The "{0}" command has been installed, type "help {0}"'
+            'for detailed help.'.format(cls.program)
```
Not sure how this is two different types of string interpolation?  I like this one better than f strings because then I'd have to repeat the variable twice, where as this is clear I'm using it in two places.

bulbazord replied:
306 - 309 is using the printf-style string interpolation and 312-314 is using the `.format` methodology.
I personally prefer f-strings, but I don't have a compelling technical reason. I do think there is value in consistency though.

jimingham commented on lldb/include/lldb/Interpreter/ScriptInterpreter.h:
```diff
@@ -476,6 +476,14 @@ class ScriptInterpreter : public PluginInterface {
+  virtual bool RunScriptBasedParsedCommand(
+      StructuredData::GenericSP impl_obj_sp, Args& args,
+      ScriptedCommandSynchronicity synchronicity,
+      lldb_private::CommandReturnObject &cmd_retobj, Status &error,
```
This is another place where I'm just following what the Raw command form did.  We can go clean up both, and probably should, but that seems like a separate pass over the code.

bulbazord replied:
Agreed.

jimingham commented on lldb/source/Commands/CommandObjectCommands.cpp:
```diff
@@ -1255,6 +1258,676 @@ class CommandObjectScriptingObject : public CommandObjectRaw {
+      return llvm::ArrayRef(m_options_definition_up.get(), m_num_options);
+    }
+    
+    static bool ParseUsageMaskFromArray(StructuredData::ObjectSP obj_sp, 
+        size_t counter, uint32_t &usage_mask, Status &error) {
```
Old habits

jimingham commented:
I addressed most of the review comments - a few I had questions about - and also did the clang-format & python-deformatting.

bulbazord commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,315 @@
+            (value, error)  = __class__.translate_value(elem["value_type"], opt_value)
+
+        if not error:
+            object.__setattr__(self, elem["varname"], value)
```
I'm primarily going off of the documentation for `AttributeError` -- https://docs.python.org/3/library/exceptions.html#AttributeError
Specifically it says that this gets raised when "an attribute reference (...) or assignment fails."
Note that you already do this on line 190. If we don't need to explicitly catch `AttributeError` here then we should be consistent and remove that one too.

jimingham replied:
That's done for a different purpose.  

I allow you to have a "default" entry for any given option, and also a "varname" for it.  If both those are provided, then this code initializes the python property for this option back to its default in the options dict.  But if you had a fancy initialization that didn't have a constant default, you can leave the default out, override `option_parsing_starting` and do the initialization yourself.
So the `try` here is just because I am not requiring there be a default/varname pair for every option and this is a cheap way to pass over the ones that don't have that pair.  I could also do explicit checks for whether the two fields exist, but this seems straightforward.

jimingham committed:
First pass at addressing review comments

jimingham commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,315 @@
+lldb parsed commands more Pythonic.
+The way to use it is to make a class for you command that inherits from ParsedCommandBase.
+That will make an LLDBOVParser which you will use for your
+option definition, and to fetch option values for the current invocation
```
I don't think your version is grammatical.  You use it (a) for your... and (b) to do X.  "for" doesn't work with the fetch verb form.  You could say "for your option definition and fetching"...
But I actually don't see what's unclear about this form.

DavidSpickett replied:
I think it's clear enough as is.

bulbazord replied:
I don't remember why I suggested this, it looks pretty clear to me after another pass.

jimingham commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,315 @@
+            return lldb.eNoCompletion
+
+    def get_option_element(self, long_name):
+        # Fixme: Is it worth making a long_option dict holding the rest of
+        # the options dict so this lookup is faster?
```
Yup, that's better.

jimingham committed:
Address more review comments

jimingham committed:
Finishing up the current review comments in CommandObjectCommands.cpp.

jimingham committed:
Convert the cmdtemplate.py to use the lldb parsed commands.
Also remove a couple stray prints in parsed_cmd.py.

jimingham committed:
Change usage->help and varname->dest

No reason to have unnecessary differences from the way argparse
did things.  This will make it easier to convert.  `help` however is
not an optional argument in the new setup.

jimingham commented:
Okay, I made one more trivial change, I switched the `varname` field to `dest` so it matched what argparse had, and I renamed the `usage` entry to `help` so it matches the argparse usage as well.

jimingham commented:
At this point, I'd like to get this in so I can iterate on it in future PR's rather than churning this one.  If folks have a bit to look over this, that would be great.  Thanks!

DavidSpickett commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,313 @@
+You can access the option values using the 'dest' string you passed in when defining the option.
+
+If you need to know whether a given option was set by the user or not, you can
+use the was_set API.
```
From what, is this an SB API thing or some python builtin?

DavidSpickett replied:
If you have a one line example like the ones above that'd be good.

jimingham replied:
Yes, it's a part of the ov_parser API.  I added examples.

DavidSpickett commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,313 @@
+import sys
+from abc import abstractmethod
+
+class LLDBOVParser:
```
I think I said this before, but I would just expand the class name to `LLDBOptionValueParser` (despite how Java enterprise ready (tm) that sounds).

It'll be read many more times than it's written and this example file is already enough to think about without chasing down what this acronym means. If the user's code is going to be littered with the literal name, then maybe keep it short but otherwise it's just one more unknown for them.

DavidSpickett replied:
`self.ov_parser` is fine as a variable name since any IDE helper will show this type name for it.

jimingham replied:
Yes, I agree, nobody will ever type this one because that's not even the class you override.  So there's no reason for it to have a mysterious name.

DavidSpickett commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,313 @@
+            value = True
+            error = False
+            
+        if not value and low_in == "n" or low_in == "no" or low_in == "f" or low_in == "false" or low_in == "0":
```
`if low_in in [<all of the possible things>]:` is slightly easier to read and update if needed.

Also error is always set false so move that out of the ifs to before the return.

jimingham replied:
string in array of strings is definitely easier to read, thanks!

I don't think the second comment is right, however.  Not all input values will make it into either of those if statements, for instance "low_in = definitely_not_a_boolean" will be rejected by both if's and so "error" will be left at True.

All of this conversion code should actually go away, because it's silly to have two versions of OptionArgParser::ToBool, one in C++ and one here in Python.  This is just a shortcut for now because this patch was already getting somewhat unwieldily.

DavidSpickett commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,313 @@
+        try:
+            return cls.completion_table[arg_type]
+        except KeyError:
+            return lldb.eNoCompletion
```
Could be `cls.completion_table.get(arg_type, lldb.eNoCompletion)`.

jimingham replied:
Yes, apparently I learned about get between writing this method and the next one...

DavidSpickett commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,313 @@
+            try:
+                object.__setattr__(self, elem["dest"], elem["default"])
+            except AttributeError:
+            # It isn't an error not to have a target, you'll just have to set and
+            # get this option value on your own.
```
Indent to match continue.

DavidSpickett replied:
Also what is "target" here? Is that the option you want to set the value on?

jimingham replied:
Right, I think I made the comment a little clearer.  I don't want to go too much into this right now.  One way to extend these commands for fancier option setting is to override `set_option_value` and handle your special options first, then delegate the others to the base class method.  But I think it would be more convenient to add a way to do custom setters.  Anyway, I'm not decided on that yet.

DavidSpickett commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,313 @@
+        else:
+            (value, error)  = __class__.translate_value(elem["value_type"], opt_value)
+
+        if not error:
+            object.__setattr__(self, elem["dest"], value)
+            elem["_value_set"] = True
+            return True
+        return False
```
```
if error:
  return False

object.__setattr__(self, elem["dest"], value)
elem["_value_set"] = True
return True
```
Seems simpler to me.

jimingham replied:
Sure.

jimingham committed:
Address David's latest review comments.

bulbazord reviewed (approved):
Ok, I think this is good to go in. There are things that can be improved but it would be better to land this and change things in follow up PRs instead of continually adjusting this one.

medismailben commented on lldb/source/Commands/Options.td:
```diff
@@ -805,19 +805,25 @@ let Command = "script add" in {
     Desc<"Name of the Python function to bind to this command name.">;
-  def script_add_class : Option<"class", "c">, Group<2>, Arg<"PythonClass">,
-  Desc<"Name of the Python class to bind to this command name.">;
+  def script_add_class : Option<"class", "c">, Groups<[2,3]>, 
```
Let's swap the short-option between `script_add_class` & `script_add_completion_type`. Scripted Process uses `-C` (capital C) in `process launch` & `process attach` to specify the python class. Same for scripted thread plans with `thread step-scripted`.

Let's try to stay consistent for our users.

jimingham replied:
I didn't add either of these options, they predate this patch.  This would break anybody that's adding scripted commands.  I don't remember why it's this way, but I don't think we can change it.

medismailben commented on lldb/source/Commands/Options.td:
```diff
@@ -805,19 +805,25 @@ let Command = "script add" in {
-  def completion_type : Option<"completion-type", "C">,
-  EnumArg<"CompletionType">,
-  Desc<"Specify which completion type the command should use - if none is specified, the command won't use auto-completion.">;
+  def script_add_completion_type : Option<"completion-type", "C">, 
```
This one also needs to be swapped.

jimingham replied:
See above.

medismailben commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,321 @@
+    def add_argument_set(self, arguments):
+        self.args_array.append(arguments)
+
+class ParsedCommandBase:
```
If this is supposed to be the base implementation, I think it needs to move to https://github.com/llvm/llvm-project/tree/main/lldb/examples/python/templates in its own file.

Also, I'd drop the `Base` suffix on the class name, since we don't do this for other base implementation classes.

jimingham replied:
Sure

medismailben reviewed (changes requested):
The implementation LGTM but as you know, I'm trying to consolidate scripting affordances & improve their documentation, so I left some comments that would greatly help me in this endeavor. Thanks!

jimingham commented on lldb/bindings/python/CMakeLists.txt:
```diff
@@ -96,7 +96,9 @@ function(finish_swig_python swig_target lldb_python_bindings_dir lldb_python_tar
     "utils"
     FILES "${LLDB_SOURCE_DIR}/examples/python/in_call_stack.py"
-          "${LLDB_SOURCE_DIR}/examples/python/symbolication.py")
+          "${LLDB_SOURCE_DIR}/examples/python/parsed_cmd.py"
```
Sure.  I moved the `parsed_cmd.py` file into lldb/examples/python/templates as well, since the other files were there.  TTTT, that seems like a strange location for these files.  At this point they really aren't examples, they are base classes we're vending.  That's also true of things like crashlog, however, so I don't feel the need to address this right now.

jimingham committed:
Address Ismail's review comments.

jimingham commented on lldb/examples/python/parsed_cmd.py:
```diff
@@ -0,0 +1,321 @@
+There are example commands in the lldb testsuite at:
+
+llvm-project/lldb/test/API/commands/command/script/add/test_commands.py
+"""
```
I added some docs, see what you think.

medismailben reviewed (approved):
Awesome! Thanks for addressing my comments @jimingham :) LGTM!

medismailben commented on lldb/examples/python/templates/parsed_cmd.py:
```diff
@@ -0,0 +1,360 @@
+
+class LLDBOptionValueParser:
+    """
+    This class is holds the option definitions for the command, and when
```
nit: typo

```suggestion
    This class holds the option definitions for the command, and when
```

jimingham committed:
One more review nit.
